tools/linter/discovery.rb

```diff
@@ -0,0 +1,107 @@
+# Copyright 2018 Google Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'net/http'
+require 'json'
+require 'active_support/inflector'
+
+# Holds all Discovery Information about a Api::Type (property)
+class DiscoveryProperty
+  attr_reader :schema
+  attr_reader :name
+
+  attr_reader :__product
+
+  def initialize(name, schema, product)
+    @name = name
+    @schema = schema
+
+    @__product = product
+  end
+
+  def nested_properties?
+    !nested_properties.empty?
+  end
+
+  def nested_properties
+    if @schema.dig('$ref')
+      @__product.get_resource(@schema.dig('$ref')).properties
+    elsif @schema.dig('type') == 'object' && @schema.dig('properties')
+      DiscoveryResource.new(@schema, nil, @__product).properties
+    elsif @schema.dig('type') == 'array' && @schema.dig('items', '$ref')
+      @__product.get_resource(@schema.dig('items', '$ref')).properties
+    elsif @schema.dig('type') == 'array' && @schema.dig('items', 'properties')
+      DiscoveryResource.new(@schema.dig('items'), nil, @__product).properties
+    else
+      []
+    end
+  end
+end
+
+# Holds Discovery information about a Resource
+# This Resource is usually a Api::Resource,
```

Comment:
I didn't want to call it a schema (which is what it really is) because eventually we'll be grabbing method information and adding it to these resources.